rust/lance-index/src/scalar/ngram.rs

```diff
@@ -0,0 +1,456 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::any::Any;
+use std::{collections::HashMap, sync::Arc};
+
+use arrow::array::AsArray;
+use arrow::datatypes::UInt64Type;
+use arrow_array::{BinaryArray, RecordBatch, StringArray};
+use arrow_schema::{DataType, Field, Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::execution::SendableRecordBatchStream;
+use deepsize::DeepSizeOf;
+use futures::{StreamExt, TryStreamExt};
+use lance_core::utils::address::RowAddress;
+use lance_core::Result;
+use lance_core::{utils::mask::RowIdTreeMap, Error};
+use moka::future::Cache;
+use roaring::{RoaringBitmap, RoaringTreemap};
+use serde::Serialize;
+use snafu::location;
+use tantivy::tokenizer::TextAnalyzer;
+use tracing::instrument;
+
+use crate::scalar::inverted::CACHE_SIZE;
+use crate::vector::VectorIndex;
+use crate::{Index, IndexType};
+
+use super::btree::TrainingSource;
+use super::inverted::TokenSet;
+use super::{AnyQuery, IndexReader, IndexStore, ScalarIndex, SearchResult, TextQuery};
+
+const TOKENS_COL: &str = "tokens";
+const POSTING_LIST_COL: &str = "posting_list";
+const POSTINGS_FILENAME: &str = "ngram_postings.lance";
+
+lazy_static::lazy_static! {
+    pub static ref TOKENS_FIELD: Field = Field::new(TOKENS_COL, DataType::Utf8, false);
+    pub static ref POSTINGS_FIELD: Field = Field::new(POSTING_LIST_COL, DataType::Binary, false);
+    pub static ref POSTINGS_SCHEMA: SchemaRef = Arc::new(Schema::new(vec![TOKENS_FIELD.clone(), POSTINGS_FIELD.clone()]));
+    /// Currently we ALWAYS use trigrams with ascii folding and lower casing.  We may want to make this configurable in the future.
```

Comment:
what's the benefit of ascii folding? Won't that cause a lower rate of success on recheck?

Reply:
I suppose I was thinking we might someday want an `ascii_folding_contains(prompt, cafe)` but I suppose that would mean we'd need to write the equivalent UDF for the in-memory check.

Reply:
I was wondering if it might be costing us time in the build as well